### PR TITLE
Feature/reg 1234 sample validation

### PR DESCRIPTION
**Pull request by ameenhaq**
* Created validation types; Success and Failure for validation check result types
* Map function to convert Validation.types to Option
* Function to validate prn-startpoint and log if invalid
* Function to validate sample size and log if invalid

### CODE REVIEW

File: resources/features/Sampling.feature [hunk 1]
```diff
@@ -0,0 +1,106 @@
+Feature: Making a Sample
+    The sample creation method will ingest a frame and a set of stratification parameters to produce an output sample.
+    To do this, the pre-filtered dataset is sorted in ascending order by 'prn'. Each of the stratification stratas with
```
Comment (pwetherill) [inline resources/features/Sampling.feature]: Question - are you sorting the total input frame by PRN prior to actioning the individual strata?
It's not a problem if this allows the creation of each strata in PRN order themselves. It's a different approach to our current system which allocates to strata first and then sorts each strata by PRN.

File: resources/features/Sampling.feature [hunk 2]
```diff
@@ -0,0 +1,106 @@
...
+    a selection type parameter set as 'Census' or 'Prn Sampling' (ommitting 'Universal') is then processed. We then use
+    the start point parameter (of a given strata), and select units in prn order until the required sample size is
```
Comment (pwetherill) [inline resources/features/Sampling.feature]: There is no requirement to use PRN points for a 'Census' cell as all units are selected. The stratification parameters hold a dummy PRN value of 0.000000000 and number required value of 0 for the 'Census' cells to reflect this.
For clarity we should state that where the selection type parameter is set to 'Census', all units are selected; where the type is 'PRN Sampling' we use the start point and select units in PRN order until the number required is reached.

File: resources/features/Sampling.feature [hunk 3]
```diff
@@ -0,0 +1,106 @@
...
+    reached. All samples are outputted collectively as a DataFrame, with an appended 'Cell Number' to each to indicate
+    the strata to which it belongs.
+
```
Comment (pwetherill) [inline resources/features/Sampling.feature]: The scenarios seem to imply that there will be separate data frames by selection type, however the requirement is for a collated output i.e. a single data frame containing all 'Census' and 'PRN' sampled units + exported as csv

File: resources/features/Sampling.feature (hunk 1 above)
Comment (ameenhaq) [reply resources/features/Sampling.feature]: Yes

File: resources/features/Sampling.feature (hunk 2 above)
Comment (ameenhaq) [reply resources/features/Sampling.feature]: Good point - seems I didn't mention a critical part

File: resources/features/Sampling.feature (hunk 3 above)
Comment (ameenhaq) [reply resources/features/Sampling.feature]: Maybe I should reword this. Currently for each strata a sample is created and appends this with the given CellNumber from the strata and then the next strata is appended to the existing strata DataFrame until all strata are either processed or some omitted.

File: resources/features/Sampling.feature [hunk 4]
```diff
@@ -0,0 +1,106 @@
...
+    Scenario Outline: A strata for Census is created
+        Given a frame and a stratification properties file paths:
+            | data_input_path   | strat_properties        | expected_output   |
+            | <input_path>      | <strat_properties_path> | <output_path>     |
+        When a <language> Sample is created from the pre-filtered frame
+        Then a Sample DataFrame containing the Census strata is returned and exported to CSV
```
Comment (nigelhp) [inline resources/features/Sampling.feature]: How can the business use this test to confirm that stratification has been applied correctly?

I appreciate that a certain style has been adopted by the existing Cucumber tests - but generally assertions that are hidden behind general steps such as "then a sample dataFrame is returned" are not the spirit of Cucumber.  The business has no way of confirming the step.

The preferred approach is for the Then step to define a table that contains the expected output.  The business then has input into the content of that table - and has all the information it needs in the test itself to confirm that the code has done exactly what they wanted it to.

File: src/test/scala/uk/gov/ons/registers/stepdefs/CommonSteps.scala [hunk 5]
```diff
@@ -0,0 +1,30 @@
...
+
+  Given("a frame and a stratification properties file paths:$") { dataTable: DataTable =>
+    import ParsingUtils._
+
+    val dataTableAsListOfMaps = dataTable.asMaps(classOf[String], classOf[String])
+    val mapOfPaths = retrievePath(dataTableAsListOfMaps) _
+    inputPath = mapOfPaths(InputPath)
+    outputPath = mapOfPaths(OutputPath)
+    stratificationPropertiesPath = mapOfPaths(StratificationPropertiesPath)
+    println(s"Given file paths for inputPath [$inputPath], outputPath [$outputPath] and stratificationPropertiesPath [$stratificationPropertiesPath]")
```
[the remark this answers is not in the document]
Comment (ameenhaq) [reply src/test/scala/uk/gov/ons/registers/stepdefs/CommonSteps.scala]: Will do when logging approach is implemented.

File: src/main/scala/uk/gov/ons/registers/ParamValidation.scala [hunk 6]
```diff
@@ -0,0 +1,34 @@
...
+      logWithErrorMsg(strataNumber)(msg = s"Error: Sample size [$sampleSize] must be a natural number less than $maxSize. " +
+        s"Parameter overridden with with max sample size [$maxSize]")
+      Success(optB = Some(maxSize)) }
+    else if (sampleSize > min && sampleSize < maxSize) Success(optB = Some(sampleSize))
+    else Failure(s"Error: Sample size [$sampleSize] must be a natural number greater than $min and less than $maxSize")
+
+
+  private def logWithErrorMsg[A](strataNumber: Int)(msg: A): Option[Nothing] = {
+    val logErrorMsg = s"[WARN] Could not process strata ($strataNumber): $msg"
+    // TODO - REPLACE AND ADD LOGGER
```
[the remark this answers is not in the document]
Comment (ameenhaq) [reply src/main/scala/uk/gov/ons/registers/ParamValidation.scala]: Logging is another story

File: src/main/scala/uk/gov/ons/registers/ParamValidation.scala [hunk 7]
```diff
@@ -0,0 +1,34 @@
...
+  private def validateSampleSize(strataNumber: Int, maxSize: Int, sampleSize: Int): Validation[Nothing, Int, ErrorMessage] =
+    if (sampleSize > maxSize) {
+      logWithErrorMsg(strataNumber)(msg = s"Error: Sample size [$sampleSize] must be a natural number less than $maxSize. " +
+        s"Parameter overridden with with max sample size [$maxSize]")
+      Success(optB = Some(maxSize)) }
+    else if (sampleSize > min && sampleSize < maxSize) Success(optB = Some(sampleSize))
+    else Failure(s"Error: Sample size [$sampleSize] must be a natural number greater than $min and less than $maxSize")
+
+
+  private def logWithErrorMsg[A](strataNumber: Int)(msg: A): Option[Nothing] = {
```
[the remark this answers is not in the document]
Comment (ameenhaq) [reply src/main/scala/uk/gov/ons/registers/ParamValidation.scala]: It in part used by the Validation.mapOption function which takes an onFailure function add has to return a type that conforms to the expected type of Validation.mapOption - i.e. Option[Int]. In this case a None as its invalid - the same segment of code is used twice also, hence I felt appropriate to pass as a function

**Pull request closed without merging**